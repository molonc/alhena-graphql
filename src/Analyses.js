const { gql } = require("apollo-server");

import client from "./api/client";

import bodybuilder from "bodybuilder";
import _ from "lodash";

const FIELD_HIERARCHY = ["project", "sample_id", "library_id", "jira_id"];
const FIELD_NAMES = {
  project: "Project",
  sample_id: "Sample ID",
  library_id: "Library ID",
  jira_id: "Jira ID"
};

export const schema = gql`
  extend type Query {
    analysesTree(filters: [Term]!): AnalysesTree
    analysesList(filters: [Term]!): [AnalysisGroup!]
    analysesStats(filters: [Term]!): [Stat!]!
  }
  input Term {
    label: String!
    value: String!
  }

  type AnalysesTree {
    children: [ParentType!]
  }
  interface NodeType {
    name: String!
  }
  type ParentType implements NodeType {
    name: String!
    children: [NodeType!]
  }
  type ChildType implements NodeType {
    name: String!
    value: Int!
  }

  type AnalysisGroup {
    label: String!
    type: String!
    values: [String!]
  }

  type Stat {
    label: String!
    value: Int!
  }
`;

const filterChildren = (root, hierarchyLevel) => {
  const uniqueRoot = _.uniq(
    root.filtered.map(analysis => analysis[FIELD_HIERARCHY[hierarchyLevel]])
  );

  const mappedRoot = uniqueRoot.map(field => ({
    name: field,
    hierarchyLevel: hierarchyLevel + 1,
    filtered: root.filtered.filter(
      analysis =>
        analysis[FIELD_HIERARCHY[hierarchyLevel]].localeCompare(field) === 0
    )
  }));

  return mappedRoot;
};

async function getAnalyses(filters) {
  const baseQuery = bodybuilder().size(10000);

  const query =
    filters === []
      ? baseQuery.build()
      : filters
          .reduce(
            (query, filter) =>
              query.filter("term", filter["label"], filter["value"]),
            baseQuery
          )
          .build();

  const data = await client.search({
    index: "analyses",
    body: query
  });
  return data["body"]["hits"]["hits"].map(hit => hit["_source"]);
}

const getUniqueValuesInKey = (list, key) =>
  list
    .map(element => element[key])
    .reduce(
      (uniques, value) =>
        uniques.indexOf(value) === -1 ? [...uniques, value] : uniques,
      []
    );

export const resolvers = {
  NodeType: {
    name: root => root.name,
    __resolveType(event, context, info) {
      if (event.hierarchyLevel === FIELD_HIERARCHY.length) {
        return "ChildType";
      } else {
        return "ParentType";
      }
    }
  },
  ParentType: {
    children: root => filterChildren(root, root.hierarchyLevel)
  },
  ChildType: {
    value: () => 1
  },
  AnalysesTree: {
    children: root => filterChildren({ filtered: [...root] }, 0)
  },

  AnalysisGroup: {
    label: root => root.label,
    type: root => root.type,
    values: root => root.values
  },
  Stat: {
    label: root => root.label,
    value: root => root.value
  },
  Query: {
    analysesTree: async (_, { filters }) => {
      const data = await getAnalyses(filters);
      return data;
    },

    analysesList: async (_, { filters }) => {
      const data = await getAnalyses(filters);

      const uniqueValuesInHierarchy = FIELD_HIERARCHY.map(field => {
        const values = getUniqueValuesInKey(data, field);

        return {
          label: FIELD_NAMES[field],
          type: field,
          values: values
        };
      });

      return uniqueValuesInHierarchy;
    },

    analysesStats: async (_, { filters }) => {
      const data = await getAnalyses(filters);

      // Return count of each thing in the hierarchy

      const counts = FIELD_HIERARCHY.map(field => {
        const values = getUniqueValuesInKey(data, field);

        return {
          label: FIELD_NAMES[field],
          value: values.length
        };
      });

      return counts;
    }
  }
};
