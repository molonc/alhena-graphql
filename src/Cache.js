const { gql } = require("apollo-server");
import redis from "./api/redisClient.js";

import _ from "lodash";

import cacheConfig from "./api/cacheConfigs.js";

import bodybuilder from "bodybuilder";
export const schema = gql`
  extend type Query {
    setCache(type: String!, value: Int!, auth: ApiUser!): Confirmation
  }
`;

export const resolvers = {
  Query: {
    async setCache(_, { type, value, auth }) {
      await redis.set(cacheConfig[type] + auth.uid, value);
      return true;
    }
  }
};
