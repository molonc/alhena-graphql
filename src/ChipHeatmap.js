const { gql } = require("apollo-server");

import _ from "lodash";
import client from "./api/montageClient.js";

export const schema = gql`
  extend type Query {
    chipHeatmap(analysis: String, quality: String!): ChipHeatmap
  }
  type ChipHeatmap {
    squares: [Square]
    stats: ChipStats
  }
  type ChipStats {
    max: Int
  }
  type Square {
    columnIndex: Int
    rowIndex: Int
    cellId: String
    heatmapOrder: Int
    totalMappedReads: String
  }
`;

export const resolvers = {
  Query: {
    async chipHeatmap(_, { analysis, quality }) {
      return await getChipHeatmap(analysis, quality);
    }
  },
  ChipHeatmap: {
    stats: root => root.max_mapped_reads,
    squares: root =>
      root.column.buckets
        .map(column => {
          const key = column.key;
          return column.row.buckets.map(rowEntry => ({
            colIndex: key,
            ...rowEntry
          }));
        })
        .flat(1)
  },
  ChipStats: {
    max: root => root.value
  },
  Square: {
    columnIndex: root => root.colIndex,
    rowIndex: root => root.key,
    heatmapOrder: root =>
      root.heatmap_order.buckets.length > 0
        ? root.heatmap_order.buckets[0].key
        : null,
    cellId: root => root.cell_id.buckets[0].key,
    totalMappedReads: root => root.total_mapped_reads.value
  }
};
const getChipHeatmap = async (analysis, quality) => {
  const results = await client.search({
    index: analysis.toLowerCase(),
    size: 0,
    body: {
      aggs: {
        max_mapped_reads: { max: { field: "total_mapped_reads" } },
        column: {
          terms: {
            size: 1000,
            field: "column",
            order: {
              _term: "asc"
            }
          },
          aggs: {
            row: {
              terms: {
                size: 1000,
                field: "row",
                order: {
                  _term: "asc"
                }
              },
              aggs: {
                cell_id: {
                  terms: {
                    size: 1000,
                    field: "cell_id"
                  }
                },
                heatmap_order: {
                  terms: {
                    size: 1000,
                    field: "all_heatmap_order"
                  }
                },
                total_mapped_reads: {
                  avg: {
                    field: "total_mapped_reads"
                  }
                },
                experimental_condition: {
                  terms: {
                    size: 1000,
                    field: "experimental_condition"
                  }
                }
              }
            }
          }
        }
      },
      query: {
        filtered: {
          filter: {
            bool: {
              must: [
                {
                  terms: {
                    caller: ["single_cell_qc"]
                  }
                },
                {
                  range: {
                    quality: {
                      gte: quality
                    }
                  }
                }
              ]
            }
          },
          query: {
            match_all: {}
          }
        }
      }
    }
  });
  return results.body.aggregations;
};
