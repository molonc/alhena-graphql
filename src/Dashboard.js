const {gql} = require("apollo-server");
import {getAllDashboards} from "./api/client.js";

export const schema = gql`
  interface Dashboard {
    index: String!
    libraryID: String!
    project: String!
    sampleID: String!
  }
  extend type Query {
    getAllDashboards: [Dashboard!]
  }
  type QCDashboard implements Dashboard {
    index: String!
    libraryID: String!
    project: String!
    sampleID: String!
  }
  type FitnessDashboard implements Dashboard {
    index: String!
    libraryID: String!
    project: String!
    sampleID: String!
  }
  type SpectrumDashboard implements Dashboard {
    index: String!
    libraryID: String!
    project: String!
    sampleID: String!
  }
`;

export const resolvers = {
  Dashboard: {
    index: root => root._source.jira_id,
    libraryID: root => root._source.library_id,
    project: root => root._source.project,
    sampleID: root => root._source.sample_id,
    __resolveType(event, context, info) {
      if (event._source.project.indexOf("DLP") !== -1) {
        return "QCDashboard";
      } else if (event._source.project.indexOf("Spectrum") !== -1) {
        return "SpectrumDashboard";
      } else {
        return "FitnessDashboard";
      }
    }
  },
  Query: {
    getAllDashboards: async () => {
      return await getAllDashboards();
    }
  }
};
