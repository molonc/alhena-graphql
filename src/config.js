export const configConsts = [
  {
    label: "Autocorrelation Hmmcopy",
    type: "autocorrelation_hmmcopy",
    class: "numerical"
  },

  { label: "Breakpoints", type: "breakpoints", class: "numerical" },
  {
    label: "Breadth Coverage Depth",
    type: "coverage_breadth",
    class: "numerical"
  },
  { label: "Cell Call", type: "cell_call", class: "categorical" },
  { label: "Column", type: "column", class: "categorical" },
  { label: "Coverage Depth", type: "coverage_depth", class: "numerical" },

  { label: "Cv Hmmcopy", type: "cv_hmmcopy", class: "numerical" },

  {
    label: "Empty Bins Hmmcopy",
    type: "empty_bins_hmmcopy",
    class: "numerical"
  },
  {
    label: "Estimated Library Size",
    type: "estimated_library_size",
    class: "numerical"
  },
  {
    label: "Experimental Condition",
    type: "experimental_condition",
    class: "categorical"
  },

  {
    label: "Fastqscreen Grch37",
    type: "fastqscreen_grch37",
    class: "numerical"
  },
  {
    label: "Fastqscreen Grch37 Multihit",
    type: "fastqscreen_grch37_multihit",
    class: "numerical"
  },
  { label: "Fastqscreen Mm10", type: "fastqscreen_mm10", class: "numerical" },
  {
    label: "Fastqscreen Mm10 Multihit",
    type: "fastqscreen_mm10_multihit",
    class: "numerical"
  },
  { label: "Fastqscreen Nohit", type: "fastqscreen_nohit", class: "numerical" },
  {
    label: "Fastqscreen Salmon",
    type: "fastqscreen_salmon",
    class: "numerical"
  },
  {
    label: "Fastqscreen Salmon Multihit",
    type: "fastqscreen_salmon_multihit",
    class: "numerical"
  },
  {
    label: "Is S Phase Probability",
    type: "is_s_phase_prob",
    class: "numerical"
  },
  { label: "Log Likelihood", type: "log_likelihood", class: "numerical" },
  { label: "MBRSI Dispersion", type: "MBRSM_dispersion", class: "numerical" },
  {
    label: "MBRSI Dispersion Non Integerness",
    type: "MBRSI_dispersion_non_integerness",
    class: "numerical"
  },
  {
    label: "MBRSI Non Integerness",
    type: "MBRSI_non_integerness",
    class: "numerical"
  },
  { label: "Mad Hmmcopy", type: "mad_hmmcopy", class: "numerical" },
  { label: "Mad Neutral State", type: "mad_neutral_state", class: "numerical" },
  { label: "Mean Copy", type: "mean_copy", class: "numerical" },
  {
    label: "Mean Hmmcopy Reads Per Bin",
    type: "mean_hmmcopy_reads_per_bin",
    class: "numerical"
  },
  { label: "Mean Insert Size", type: "mean_insert_size", class: "numerical" },
  { label: "Mean State Mads", type: "mean_state_mads", class: "numerical" },
  { label: "Mean State Vars", type: "mean_state_vars", class: "numerical" },
  {
    label: "Median Insert Size",
    type: "median_insert_size",
    class: "numerical"
  },
  { label: "Multiplier", type: "multiplier", class: "numerical" },
  {
    label: "Paired Duplicate Reads",
    type: "paired_duplicate_reads",
    class: "numerical"
  },
  {
    label: "Paired Mapped Reads",
    type: "paired_mapped_reads",
    class: "numerical"
  },
  {
    label: "Percent Duplicate Reads",
    type: "percent_duplicate_reads",
    class: "numerical"
  },
  {
    label: "Percent Unmapped Reads",
    type: "percent_unmapped_reads",
    class: "numerical"
  },
  { label: "Quality", type: "quality", class: "numerical" },
  { label: "Row", type: "row" },
  { label: "Scaled Halfiness", type: "scaled_halfiness", class: "numerical" },
  {
    label: "Standard Deviation Insert Size",
    type: "standard_deviation_insert_size",
    class: "numerical"
  },
  { label: "State Mode", type: "state_mode", class: "categorical" },
  {
    label: "Std Hmmcopy Reads Per Bin",
    type: "std_hmmcopy_reads_per_bin",
    class: "numerical"
  },
  {
    label: "Total Duplicate Reads",
    type: "total_duplicate_reads",
    class: "numerical"
  },
  {
    label: "Total Mapped Reads",
    type: "total_mapped_reads",
    class: "numerical"
  },
  {
    label: "Total Mapped Reads HMMCopy",
    type: "total_mapped_reads_hmmcopy",
    class: "numerical"
  },
  { label: "Unmapped Reads", type: "unmapped_reads", class: "numerical" },
  {
    label: "Total Properly Paired",
    type: "total_properly_paired",
    class: "numerical"
  },
  { label: "Total Reads", type: "total_reads", class: "numerical" },
  { label: "True Multiplier", type: "true_multiplier", class: "numerical" },
  {
    label: "Unpaired Duplicate Reads",
    type: "unpaired_duplicate_reads",
    class: "numerical"
  },
  {
    label: "Unpaired Mapped Reads",
    type: "unpaired_mapped_reads",
    class: "numerical"
  },
  {
    label: "Is Contaminated",
    type: "is_contaminated",
    class: "boolean"
  }
];
//  { type: "project", label: "Project" },
export const defaultDashboardColumns = [
  { type: "sample_id", label: "Sample ID" },
  { type: "library_id", label: "Library ID" },
  { type: "jira_id", label: "Jira Ticket" }
];
