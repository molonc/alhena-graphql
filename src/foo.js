const { gql } = require("apollo-server");

export const schema = gql`
  extend type Query {
    foo: [Bar!]
  }

  type Bar {
    id: String!
    name: String!
    age: Int!
  }
`;

export const resolvers = {
  Query: {
    foo() {
      return DATA;
    }
  },
  Bar: {
    id: root => root.id,
    name: root => root.name,
    age: root => root.age
  }
};

const DATA = [
  {
    id: "foo1",
    name: "Sam",
    age: 5
  },
  {
    id: "abc",
    name: "ABC",
    age: 100
  }
];
