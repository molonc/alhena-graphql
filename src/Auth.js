import redis from "./api/redisClient.js";
const { gql, AuthenticationError } = require("apollo-server");

import authClient from "./api/authClient";
import client from "./api/client";

import { superUserRoles } from "./api/utils/config.js";
import { createSuperUserClient, getRedisApiKey } from "./utils.js";

import _ from "lodash";

export const schema = gql`
  extend type Query {
    login(user: User!): LoginAcknowledgement
    getUsers(auth: ApiUser!): [AppUsers]
    createNewUser(user: NewUser!): CreationAcknowledgement
    verifyNewUserUri(key: String!): NewUserAcknowledgement
    updateUserRoles(
      newRoles: [String!]
      username: String!
      email: String!
      name: String!
    ): CreationAcknowledgement
    deleteUser(username: String!): DeletionAcknowledgement
  }
  input NewUser {
    username: String!
    email: String!
    name: String!
    password: String!
  }
  input AuthKey {
    key: String!
  }
  input User {
    uid: String!
    password: String!
  }
  input ApiUser {
    uid: String!
    authKeyID: String!
  }
  type AppUsers {
    username: String
    roles: [String]
    full_name: String
    email: String
    enabled: String
  }
  type NewUserAcknowledgement {
    isValid: Boolean!
    email: String
  }
  type DeletionAcknowledgement {
    isDeleted: Boolean
  }
  type CreationAcknowledgement {
    created: Boolean
  }
  type LoginAcknowledgement {
    statusCode: String
    authKeyID: String
    role: [String]
  }
`;

const verifyUriKey = async key => await redis.get(key);

const deleteUser = async username => {
  var response = await authClient(
    process.env.ES_USER,
    process.env.ES_PASSWORD
  ).security.deleteUser({
    username: username,
    refresh: "wait_for"
  });
  return response.body;
};

const createNewUser = async user => {
  var roles = await redis.get("roles_" + user.email);

  var response = await authClient(
    process.env.ES_USER,
    process.env.ES_PASSWORD
  ).security.putUser({
    username: user.username,
    refresh: "wait_for",
    body: {
      password: user.password,
      full_name: user.name,
      email: user.email,
      roles: [...roles.split(",")]
    }
  });

  return response.body;
};
const getUsers = async auth => {
  const authKey = await redis.get(auth.uid + ":" + auth.authKeyID);
  const data = await client(authKey, auth.authKeyID).security.getUser({});
  var users =
    data.statusCode === 200
      ? Object.keys(data.body)
          .filter(
            name => !superUserRoles.hasOwnProperty(data.body[name].roles[0])
          )
          .map(name => {
            return { ...data.body[name] };
          })
          .map(user => {
            user["roles"] = user.roles.map(role => role.split("_")[0]);
            return user;
          })
      : [];
  return users;
};
const login = async user => {
  /*  const isPasswordCorrect = await authClient(user.uid, user.password).search(
    {
      index: "analyses"
    },
    function(error) {
      console.log(error);
    }
  );*/
  //console.log(isPasswordCorrect);
  const client = authClient(process.env.ES_USER, process.env.ES_PASSWORD);

  const result = await client.security.getApiKey({
    username: user.uid
  });
  //wrong auth
  if (result.statusCode === 200) {
    //if keys invalidate
    var oldKey;
    if (result.body.api_keys.length !== 0) {
      oldKey = await client.security.invalidateApiKey({
        body: { name: "login-" + user.uid }
      });
    }
    if ((oldKey && oldKey.statusCode === 200) || oldKey === undefined) {
      const newKey = await client.security.createApiKey({
        body: { name: "login-" + user.uid, expiration: "1d" }
      });
      const roleMapping = await client.security.getUser({
        username: user.uid
      });

      //store in local sotrage to expire tomorrow
      redis.set(user.uid + ":" + newKey.body.id, newKey.body.api_key);

      redis.expireat(
        user.uid + ":" + newKey.body.id,
        parseInt(+new Date() / 1000) + 86400
      );

      return {
        statusCode: newKey.statusCode,
        authKeyID: newKey.body ? newKey.body.id : null,
        role: roleMapping.body[user.uid].roles
      };
    } else {
      return { statusCode: oldKey.statusCode, authKeyID: null, role: [] };
    }
  }
};
const updateRoles = async (newRoles, username, email, name) => {
  var response = await authClient(
    process.env.ES_USER,
    process.env.ES_PASSWORD
  ).security.putUser({
    username: username,
    refresh: "wait_for",
    body: {
      email: email,
      full_name: name,
      roles: [
        ...newRoles.map(role =>
          role.indexOf("_dashboardReader") === -1
            ? role + "_dashboardReader"
            : role
        )
      ]
    }
  });
  return response.body;
};
export const resolvers = {
  Query: {
    login: async (_, { user }) => {
      return await login(user);
    },
    getUsers: async (_, { auth }) => {
      return await getUsers(auth);
    },
    createNewUser: async (_, { user }) => {
      return await createNewUser(user);
    },
    verifyNewUserUri: async (_, { key }) => {
      return await verifyUriKey(key);
    },
    deleteUser: async (_, { username }) => {
      return await deleteUser(username);
    },
    updateUserRoles: async (_, { newRoles, username, email, name }) => {
      return await updateRoles(newRoles, username, email, name);
    }
  },
  AppUsers: {
    username: ({ username }) => username,
    roles: ({ roles }) => roles,
    full_name: ({ full_name }) => full_name,
    enabled: ({ enabled }) => enabled.toString(),
    email: ({ email }) => email
  },
  NewUserAcknowledgement: {
    isValid: root => (root ? true : false),
    email: root => root
  },
  DeletionAcknowledgement: {
    isDeleted: root => root.found
  },
  CreationAcknowledgement: { created: root => root.created },
  LoginAcknowledgement: {
    statusCode: root => root.statusCode,
    authKeyID: root => root.authKeyID,
    role: root => root.role
  }
};
