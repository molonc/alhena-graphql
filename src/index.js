import "@babel/polyfill";

import { ApolloServer } from "apollo-server-express";
import { gql } from "apollo-server";

import * as foo from "./foo.js";

import { merge } from "lodash";

const baseSchema = gql`
  type Query {
    _blank: String
  }
`;

const server = new ApolloServer({
  typeDefs: [baseSchema, foo.schema],
  resolvers: merge(foo.resolvers)
});

const express = require("express");
const app = express();
server.applyMiddleware({ app });

app.listen({ port: 4000 }, () =>
  console.log(`ðŸš€ Server ready at http://localhost:4000${server.graphqlPath}`)
);
