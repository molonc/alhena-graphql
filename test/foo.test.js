import "@babel/polyfill";
import { schema, resolvers } from "../src/foo";
import { startServer } from "./test-server";
import { gql } from "apollo-server";

let query;
beforeAll(async () => {
  const client = await startServer(schema, resolvers);
  query = client.query;
});

it("has server started", () => {
  expect(query).toBeDefined();
});

it("test foo query", async () => {
  const FOO_QUERY = gql`
    query {
      foo {
        id
        name
        age
      }
    }
  `;

  const res = await query({ query: FOO_QUERY });
  expect(res).toMatchSnapshot();
});
