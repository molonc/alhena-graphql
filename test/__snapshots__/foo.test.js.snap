// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test foo query 1`] = `
Object {
  "data": Object {
    "foo": Array [
      Object {
        "age": 5,
        "id": "foo1",
        "name": "Sam",
      },
      Object {
        "age": 100,
        "id": "abc",
        "name": "ABC",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
